### PR TITLE
Allow cross compiling of native tools

### PR DESCRIPTION
**Pull request by ekrich**
The idea here is to cross compile the compiler backend for Scala Native. The front end is a plugin to the Scala Compiler (or could be built in like Scala.js for Scala 3) that would require the Scala Compiler to be cross compiled which is much harder lift because of the use of reflection and Java libraries that would require a big effort.

Why would we want to cross compile the `tools`, `nir`, and `util` project which constitutes the backend?

1. `scala-cli` could call a native application using the native artifacts directly rather than using the `scala-native-cli` project.
2. Scala Native could create an application to call as the backend which might be an option for the `sbt` and Mill plugins.
3. Scala Native could provide tighter integration with `clang` to call it directly rather that create a subprocess to call `clang`.
4. Scala Native could use other lower level LLVM tools for other unforeseen features.

Numbers 3 and 4 above would mean that a JVM version and a Native version could diverge.

So far the Sandbox works but not tests. The tools project will need the ability to run Native and JVM tests. We will also need to publish artifacts and perhaps work on better documentation.

Co-authored-by: @WojciechMazur

### CODE REVIEW

File: project/Build.scala [hunk 1]
```diff
@@ -122,28 +122,45 @@ object Build {
   lazy val junitPlugin = MultiScalaProject("junitPlugin", file("junit-plugin"))
     .settings(compilerPluginSettings)
 
-  private val withSharedSources = Def.settings(
-    // baseDirectory = project/jvm/.<scala-version>
-    Compile / unmanagedSourceDirectories += baseDirectory.value.getParentFile.getParentFile / "src"
+  private val withSharedCrossPlatformSources = Def.settings(
```
Comment (ekrich) [inline project/Build.scala:125]: Was looking at this area but didn't know exactly how to solve it. The build is quite tricky.

**Comment by ekrich** [comment]
We are failing with a missing resource file.

The file needed is here - https://github.com/scala-native/scala-native/blob/main/unit-tests/jvm/src/test/resources/BlacklistedTests.txt

```
sbt:scala-native> show  testsJVM2_12 / Test / unmanagedResourceDirectories
[info] * /Users/eric/workspace/scala-native/testsJVM/src/test/resources
[info] * /Users/eric/workspace/scala-native/shared/src/test/resources
```
The second entry comes from here, the first entry I am not sure.
https://github.com/scala-native/scala-native/blob/cb3aff0ebc3c38aa97f067ed1914a376f8399d1d/project/Settings.scala#L417

Unless we move the resource file it would have to get copied from the place listed above to:
`testsJVM/.2.12/target/scala-2.12/classes` I think.

@WojciechMazur Any help appreciated.

**Comment by ekrich** [comment]
I am also having a problem with compile and test tools:
```
> show toolsJVM2_12 / Compile /unmanagedSourceDirectories
[info] * /Users/eric/workspace/scala-native/tools/jvm/src/main/scala
[info] * /Users/eric/workspace/scala-native/tools/jvm/src/main/scala-2.12
[info] * /Users/eric/workspace/scala-native/tools/jvm/src/main/scala-2
[info] * /Users/eric/workspace/scala-native/tools/jvm/src/main/java
[info] * /Users/eric/workspace/scala-native/tools/src
```
That looks ok but this provide `scala-3` when compiling `scala-2` and visa versa.
```
> show toolsJVM2_12 / Compile /unmanagedSources
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala-3/scala/NativeCompilerTest.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala-3/scala/scalanative/IdentifiersSuite.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala-3/scala/scalanative/NIRCompilerTest3.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala-2/scala/scalanative/IdentifiersSuite.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala/scala/scalanative/LinkerSpec.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala/scala/scalanative/NIRCompilerTest.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala/scala/scalanative/OptimizerSpec.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala/scala/scalanative/NIRCompiler.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala/scala/scalanative/FilteredClassLoader.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala/scala/scalanative/FrameworkTest.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala/scala/scalanative/IncCompilationTest.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala/scala/scalanative/NativePlatform.scala
[info] * /Users/eric/workspace/scala-native/tools/src/test/scala-3/scala/scalanative/linker/StaticForwardersSuiteScala3.scala
...
```
I think it is in this area. https://github.com/scala-native/scala-native/blob/cb3aff0ebc3c38aa97f067ed1914a376f8399d1d/project/Settings.scala#L375

**Comment by armanbilge** [comment]
The history prior to the force-push is here:
https://github.com/scala-native/scala-native/commits/b7105659ac1516fab59f19bd117a2dffbd5d0f9e

**Comment by ekrich** [comment]
Should we convert the tools tests to use JUnit?

Hopefully, I rebased this ok.

**Comment by armanbilge** [comment]
> Should we convert the tools tests to use JUnit?

Yeah, that's probably a good idea. Although we should land it as a separate PR?

**Comment by WojciechMazur** [comment]
There a few unimplemented methods required by the tools. These need to be implemented/replaced to allow to actual usage of native tools: 

```Scala
[error] Found 11 missing definitions while linking
[error] Not found Top(java.io.PipedInputStream)
[error] Not found Member(Top(java.io.PipedInputStream),RE)
[error]         at https://raw.githubusercontent.com/scala/scala/v2.13.11/src/library/scala/sys/process/ProcessImpl.scala:218
[error] Not found Top(java.io.PipedOutputStream)
[error] Not found Member(Top(java.io.PipedOutputStream),RE)
[error]         at https://raw.githubusercontent.com/scala/scala/v2.13.11/src/library/scala/sys/process/ProcessImpl.scala:195
[error] Not found Top(java.nio.file.FileSystemAlreadyExistsException)
[error]         at file:/Users/wmazur/projects/scala-native/util/src/main/scala/scala/scalanative/io/VirtualDirectory.scala:159
[error] Not found Member(Top(java.nio.file.FileSystemAlreadyExistsException),RE)
[error]         at file:/Users/wmazur/projects/scala-native/javalib/src/main/scala/scala/scalanative/nio/fs/GenericFileSystemProvider.scala:23
[error] Not found Top(java.nio.file.ProviderNotFoundException)
[error] Not found Member(Top(java.nio.file.ProviderNotFoundException),RE)
[error]         at file:/Users/wmazur/projects/scala-native/javalib/src/main/scala/java/nio/file/FileSystems.scala:76
[error] Not found Top(java.security.DigestInputStream)
[error] Not found Member(Top(java.security.DigestInputStream),RL19java.io.InputStreamL27java.security.MessageDigestE)
[error]         at file:/Users/wmazur/projects/scala-native/tools/src/main/scala/scala/scalanative/build/IO.scala:176
[error] Not found Member(Top(java.util.EnumSet),D2ofL14java.lang.EnumL17java.util.EnumSetEo)
[error]         at file:/Users/wmazur/projects/scala-native/tools/src/main/scala/scala/scalanative/build/IO.scala:39
[error]         at file:/Users/wmazur/projects/scala-native/tools/src/main/scala/scala/scalanative/build/IO.scala:71
[error]         at file:/Users/wmazur/projects/scala-native/tools/src/main/scala/scala/scalanative/build/IO.scala:115
[error] Undefined definitions found in reachability phase
[error] (tools2_13 / Test / nativeLink) Undefined definitions found in reachability phase
```

- [ ] java.ioPipedInputStreams are WIP #2691 
- [ ] java.nio.file.FileSystemAlreadyExistsException 
- [ ] java.nio.file.ProviderNotFoundException 
- [ ] java.security.DigestInputStream - problematic as we don't provide valid `java.security` implementation yet
- [ ] java.util.EnumSet - missing in both Scala Native and Scala.js

Beside that for actual tests of native toolachain we need to way to integrate with Scala compiler in `test-compiler`: 
- [ ] Invoke Scala compiler with provided classpath and sources - probably by invoking process 
- [ ] Retrive list of errors from compiler - hard to due, can be done by parsing stderr, but would not be reliable. Probably instead we should move compiler specific tests to nscplugin like it is done in Scala.js. Tools tests should cover only tests requiring linking/optimizer/codegen or utils

**Comment by ekrich** [comment]
Maybe we could consider using @catap 's code inlined to create hashes. Is everything else in the build using time stamps? SHA-1 is used now but it seems we should use SHA-2 according to NIST?
https://github.com/kcrypt/scala-sha/

**Comment by catap** [comment]
If you would like I may re-license it under something useful.

**Comment by ekrich** [comment]
That is pretty much wide open for use if I understand correctly. We would just have to add some info into our License and Notice file. Are you still developing the code? - I assume it is solid.

**Comment by catap** [comment]
@ekrich this is solid.

File: tools/src/main/scala/scala/scalanative/build/NativeLib.scala [hunk 2]
```diff
@@ -182,18 +187,23 @@ private[scalanative] object NativeLib {
...
-      Files.exists(target) &&
-        Files.exists(fileshashPath) &&
-        Arrays.equals(fileshash, Files.readAllBytes(fileshashPath))
-    if (!copied) {
+    def copy() = {
       IO.deleteRecursive(target)
       IO.copyDirectory(source, target)
-      IO.write(fileshashPath, fileshash)
     }
+    if (Platform.isJVM) {
```
Comment (ekrich) [inline tools/src/main/scala/scala/scalanative/build/NativeLib.scala:194]: I thought I could work on this part using [Kirill A. Korinsky's](https://github.com/catap) work after this gets merged. Not sure about `java.security.DigestInputStream` - I think we can work around that or have a Scala Native internal one not in the java namespace.

**Pull request merged**